### ISSUE TITLE
Rule proposal: disallow objects as Map keys

### ISSUE THREAD

**Issue by antonysouthworth-halter** [typescript-eslint/typescript-eslint#10851]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [x] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [x] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [x] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

If you use something that's compared by reference like an Array as the key for a map, the map will never return anything because map keys are evaluated by `===`.

We should lint for this and disallow using reference types as map keys.

Maybe it seems super obvious to people who use Typescript a lot, but for me I would have expected TS to stop (or at least, warn) me from doing something like this.

### Fail Cases

```TypeScript
// we are changing one of our APIs which used to a sliding window size + slide
// to instead accept an enum value from a
// limited set of supported window configurations

// old API call, e.g. "24h rolling window, 6h slide"
response = api.getSmoothedTemperature(locationId, from, to, 24, 6)

// new API call, only a handful are supported
enum WindowConfiguration {
  _1h1h,
  _24h6h,
  _7d24h,
}

response = api.getSmoothedTemperature(locationId, from, to, WindowConfiguration._24h6h);

// in the backend, map from old requests to new requests, for backwards compatibility
configMap = new Map<[number, number], WindowConfiguration>([
  [[1, 1], WindowConfiguration._1h1h],
  [[24, 6], WindowConfiguration._24h6h],
  [[7 * 24, 24], WindowConfiguration._7d24h],
]);

// oops! this never works, because arrays are compared by reference!
windowConfig = configMap.get([req.size, req.slide]);
```

### Pass Cases

```TypeScript
// gross, but works, does what you want, performance probably fine for small arrays...
configMap = new Map<string, WindowConfiguration>([
  [JSON.stringify([1, 1]), WindowConfiguration._1h1h],
  [JSON.stringify([24, 6]), WindowConfiguration._24h6h],
  [JSON.stringify([7 * 24, 24]), WindowConfiguration._7d24h],
]);

windowConfig = configMap.get(JSON.stringify([req.size, req.slide]));
```

### Additional Info

Apologies if this has been discussed before! A cursory search didn't turn up anything.

**Comment by bradzacher** [comment]
> the map will never return anything because map keys are evaluated by ===.

This is objectively false!

```ts
const myObject = {};
const map = new Map([
  [myObject, 1],
]);

map.has(myObject) // true
```

Whilst it is true that you cannot pass a new object of the same shape - passing the same object is perfectly valid and really is the entire point of Map existing as a construct - same as in any language. It is up to your code to ensure that the key checks are done using stable object references.

Closing as this request is not broadly applicable, nor does it enforce a best practice in the industry, nor does it enforce something typescript specific.

**Comment by antonysouthworth-halter** [comment]
I'm not sure I totally agree that the whole point of a Map is to compare by reference; I feel like in my experience comparison by value is much more useful. But then again, I try to spend as little time in Typescript as possible 😂 so it might just be my unfamiliarity. For example, the behaviour of the following Rust program is not surprising to me:

```rust
use std::collections::HashMap;

#[derive(Hash, PartialEq, Eq)]
struct MyKey {
    a: i32,
}

pub fn main() {
    let mut map = HashMap::new();
    let myKey = MyKey { a: 1 };

    map.insert(myKey, 1);

    println!("{:?}", map.get(&MyKey { a: 1 })); // Some(1)
}
```

Regardless, I am clearly in the minority here; appreciate your reply @bradzacher!

**Comment by bradzacher** [comment]
Rust (and most other languages) don't truly do "compare by value" for their maps. Usually (as in the case of rust) it's instead "compare by hash" - I.e. You run a hash function over the value to derive a "unique" number or a string which allows for an approximation of "compare by value" without the cost of a full comparison.

JavaScript is ultimately no different - it's just that the hash function returns the memory location for objects, and the value for strings/numbers/etc. So it's a much less accurate hash function than other languages.
